### PR TITLE
Introduce ASN.1 SEQUENCE traversal API

### PR DESCRIPTION
**Pull request by hanno-arm**
__Summary:__ This PR is a spin-out from the X.509 on demand parsing work https://github.com/ARMmbed/mbedtls/pull/2478, one of the main drivers of which is to avoid dynamically allocated presentations of ASN.1 SEQUENCEs by traversing the raw ASN.1 in-place whenever needed. To this end, this PR introduces the ASN.1 API `mbedtls_asn1_traverse_sequence_of()` which can be used to traverse an ASN.1 SEQUENCE container, triggering a user-specified callback for each entry found. 
The PR also introduces two helper macros `MBEDTLS_ASN1_IS_STRING_TAG` and `MBEDTLS_OID_CMP_RAW`. Those are technically unrelated to the SEQUENCE traversal API and are included here solely to keep the number of dependencies for the on demand parsing PR low.

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
https://github.com/ARMmbed/mbed-crypto/pull/75 fixes some edge cases in the ASN.1 library and adds tests. Can you please make sure that your X.509 code and tests are compatible with these fixes? This PR doesn't have tests; I'm willing to accept it without tests but the lack of tests needs to be filed as a separate bug.

**Comment by hanno-arm** [comment]
@gilles-peskine-arm It's not fully satisfactory, but because the PR reimplements `mbedtls_asn1_get_sequence_of()` in terms of the new traversal API, any tests for the former will also test the latter.

File: include/mbedtls/asn1.h [hunk 1]
```diff
@@ -316,6 +316,58 @@ int mbedtls_asn1_get_sequence_of( unsigned char **p,
                           mbedtls_asn1_sequence *cur,
                           int tag);
 
+/**
+ * \brief                Traverse an ASN.1 SEQUENCE container and
+ *                       call a callback for each entry.
+ *
+ * \warning              This function is still experimental and may change
+ *                       at any time.
```
Comment (Patater) [inline include/mbedtls/asn1.h]: Why should this be experimental? When can it be non-experimental? Could we make asn1_internal.h and place it there if it's only for internal use for now?

File: include/mbedtls/asn1.h [hunk 2]
```diff
@@ -316,6 +316,58 @@ int mbedtls_asn1_get_sequence_of( unsigned char **p,
...
+ * \param ctx            The context to be passed to the callback \p cb.
+ *
+ * \return               \c 0 if successful the entire ASN.1 SEQUENCE
+ *                       was traversed without parsing or callback errors.
+ * \return               A negative ASN.1 error code on a parsing failure.
+ * \return               A non-zero error code forwarded from the callback
+ *                       \p cb in case the latter returns a non-zero value.
+ */
+int mbedtls_asn1_traverse_sequence_of(
+    unsigned char **p,
```
Comment (Patater) [inline include/mbedtls/asn1.h]: Can the `unsigned char` be `const unsigned char`? I don't think we'd be modifying `**p`, only `p` or `*p`.

File: include/mbedtls/asn1.h [hunk 3]
```diff
@@ -316,6 +316,58 @@ int mbedtls_asn1_get_sequence_of( unsigned char **p,
...
+    const unsigned char *end,
+    uint8_t tag_must_mask, uint8_t tag_must_val,
+    uint8_t tag_may_mask, uint8_t tag_may_val,
+    int (*cb)( void *ctx, int tag,
+               unsigned char* start, size_t len ),
```
Comment (Patater) [inline include/mbedtls/asn1.h]: Can `start` be made const?

File: library/asn1parse.c [hunk 4]
```diff
@@ -292,59 +292,56 @@ void mbedtls_asn1_sequence_free( mbedtls_asn1_sequence *seq )
...
+    cur->buf.tag = tag;
 
+    cb_ctx->cur = cur;
     return( 0 );
 }
 
+/*
+ *  Parses and splits an ASN.1 "SEQUENCE OF <tag>"
+ */
+int mbedtls_asn1_get_sequence_of( unsigned char **p,
```
Comment (Patater) [inline library/asn1parse.c]: Ah, too bad this is not const. I guess we can't have our callback with `p` const because of this.

**Comment by gilles-peskine-arm** [comment]
I rebased this pull request on top of the current base branch. The previous content by Hanno is at https://github.com/gilles-peskine-arm/mbed-crypto/tree/asn1_traversal_api-1.

Compared with the previous content:

* I removed most updates to existing documentation because they were largely superseded by #75.
* I added tests for the new functions.
* I tweaked the documentation of the new functions.

File: include/mbedtls/asn1.h (hunk 1 above)
Comment (gilles-peskine-arm) [reply include/mbedtls/asn1.h]: It isn't internal since it will be used by Mbed TLS.

File: include/mbedtls/asn1.h (hunk 2 above)
Comment (gilles-peskine-arm) [reply include/mbedtls/asn1.h]: `const unsigned char **` and `unsigned char **` do not mix. If this function used a different type, it couldn't be used with the other asn1 parsing functions.

File: include/mbedtls/asn1.h (hunk 3 above)
Comment (gilles-peskine-arm) [reply include/mbedtls/asn1.h]: I'm wondering about that. In 99% of the cases it would be const: parsing a read-only string. But it may be useful occasionally to modify an input in place.

**Comment by gilles-peskine-arm** [comment]
New CI run now that the problems with the Cypress boards are fixed: https://jenkins-internal.mbed.com/job/mbed-crypto-pr/job/PR-263-merge/20/

File: include/mbedtls/asn1.h (hunk 1 above)
Comment (hanno-arm) [reply include/mbedtls/asn1.h]: @gilles-peskine-arm @Patater What's your view on this? This function must be public, as @gilles-peskine-arm said, so what remains is to decide whether we want to declare it as experimental or not. I'm open either way. If you think the API is good as-is, I'm happy to remove the qualification.

Comment (gilles-peskine-arm) [reply include/mbedtls/asn1.h]: My preference is what you wrote: declare this function as experimental. We don't care about this function except as an internal detail of the implementation an X509 feature that we do care about. It has a complex API that I would prefer not to make official, but that I'm ok to include with the understanding that we're likely to modify it in the future.

File: include/mbedtls/asn1.h (hunk 2 above)
Comment (hanno-arm) [reply include/mbedtls/asn1.h]: I agree with @Patater in that in principle the type of `p` should be `const unsigned char **` here. However, the rest of the ASN.1 parsing library unfortunately also uses `unsigned char **`, which we can't change and which forces the use of `unsigned char **` here, too (not just for consistency, but simply because this function calls e.g. `mbedtls_asn1_get_tag()` which has a `unsigned char **` argument). An alternative would be to extend the documentation of all ASN.1 parsing functions to explicitly mention that, despite the missing `const` qualification, the `**p` is never written to, and that it is hence safe to use the functions in a `const unsigned char **` context. If we did that, we could start improving the API by using `const unsigned char **` at least for this new API function, and doing a const-cast when calling the other parsing functions.

Comment (gilles-peskine-arm) [reply include/mbedtls/asn1.h]: We can tweak the API in Mbed TLS 3. Until we get there, let's keep the API consistent.

File: include/mbedtls/asn1.h (hunk 3 above)
Comment (hanno-arm) [reply include/mbedtls/asn1.h]: I agree that it should be `const`. However, if we want to call other ASN.1 parsing functions from within the callback, then we can't have `const` here because those parsing functions don't have `const` qualification, as mentioned above.

File: include/mbedtls/asn1.h (hunk 2 above)
Comment (hanno-arm) [reply include/mbedtls/asn1.h]: Ok, no change here, then.

File: library/asn1parse.c (hunk 4 above)
Comment (hanno-arm) [reply library/asn1parse.c]: This way it would work, but what doesn't work is calling parsing functions from within `mbedtls_asn1_traverse_sequence_of()`.

**Comment by hanno-arm** [comment]
Rebase on top of `development`.

**Comment by mpg** [comment]
The CI found a problem, in `test-ref-configs`, with `config-symmetric-only.h`, the first error is:
```
/var/lib/build/include/mbedtls/asn1.h:504:5: error: unknown type name 'uint8_t'
     uint8_t tag_must_mask, uint8_t tag_must_val,
     ^
```
and the other errors look similar. I suggest including `stdint.h` directly in `asn1.h` - it probably happens to be included indirectly in some configurations including the default one, but apparently not all of them.

**Comment by hanno-arm** [comment]
Thank you @mpg, I'll fix that.

**Comment by hanno-arm** [comment]
@mpg @gilles-peskine-arm Fixed the issue by replacing `uint8_t` by `unsigned char`.

File: include/mbedtls/asn1.h (hunk 1 above)
Comment (mpg) [reply include/mbedtls/asn1.h]: The situation should be temporary, since I think we're aiming at making the whole `asn1.h` private in Mbed TLS 3.0 anyway.

**Comment by gilles-peskine-arm** [comment]
CI passes except for the Mbed OS tests which are failing for unrelated reasons.
